### PR TITLE
Benchmarks in CI.

### LINKED ISSUE #1473: Deploy infrastructure to add o1js benchmarks to CI
This can be set up using [the first benchmark](https://github.com/o1-labs/o1js/issues/923). Others can be added later. 

We require CI to run benchmarks and generate a report. In the long term we want CI to fail if performance is degraded but what does that mean exactly? What constitutes failure, do we need the ability to override? I think we should add the report but not fail the build in the first version.

Similarly, please note that a chart of historic performance may be added but is also not a requirement of this issue / first version.

### LINKED ISSUE #1487: Update benchmark runner to publish generated metrics to benchmarking backend (storage)
- InfluxDB JS/TS client configured using envars that will be provided by CI job, which in turn will be configured as repository secrets.
  - We need a bunch of additional metadata attached to the published metrics, such as execution env. name, execution env. HW config, PR name/branch, target name the metrics were sampled for, etc..
  - This metadata will serve as filtering options for Grafana dashboard.
- https://github.com/o1-labs/o1js/pull/1481

### LINKED ISSUE #1489: An ability to compare historical or base metrics with the current benchmarking results
- Example: https://github.com/boa-dev/criterion-compare-action
- [Internal discussion](https://o1-labs.slack.com/archives/C028Q27R8UC/p1709755552866189?thread_ts=1709063286.159349&cid=C028Q27R8UC).

### PR DESCRIPTION
**Pull request by shimkiv**
Closes #1473 
Closes #1487 
Closes #1489 
Closes https://github.com/o1-labs/benchmark-infra/issues/1

### CODE REVIEW

**Comment by mitschabaude** [comment]
Hey @shimkiv one ask - can we separate the core benchmarking library from the tooling to upload benchmark data?
I.e. have two library files
- `benchmark.ts` similar to original
- `benchmarking-infra.ts` (or whatever you want to call it), which depends on influx db client etc

Similarly, it would be nice if the benchmark _definitions_ like `EcdsaBenchmark` would be in a different file from the code that runs them and stores results.

Goal: I want to keep the ability to easily run these benchmarks locally and print out their results, by writing a small script that calls `Benchmark.run()`, which does not depend on and interacts with any of our infra

**Comment by shimkiv** [comment]
> Hey @shimkiv one ask - can we separate the core benchmarking library from the tooling to upload benchmark data? I.e. have two library files
> Goal: I want to keep the ability to easily run these benchmarks locally and print out their results, by writing a small script that calls `Benchmark.run()`, which does not depend on and interacts with any of our infra

Hey @mitschabaude, if you will run benchmarks locally as it was intended at first like `./run benchmarks/ecdsa.ts --bundle` then it won't talk to any of infra since it won't be configured properly.

**Comment by mitschabaude** [comment]
@shimkiv you could pass in the additional environment data like the CPU to `Benchmark.run()`, so that the benchmarking lib doesn't have to collect that data internally:

```ts
import os from "node:os";
// import benchmarks

let environment = {
  hardware: hardware: `${os.cpus()[0].model}, ${os.cpus().length} cores, ...`
  ...
 };
 
 let results = [
   ...await EcdsaBenchmark.run(environment),
   ...await SomeOtherBenchmark.run(environment),
  ];
 
 // results include measurements + environment data
 ```

**Comment by shimkiv** [comment]
> @shimkiv you could pass in the additional environment data like the CPU to `Benchmark.run()`, so that the benchmarking lib doesn't have to collect that data internally:

This is cool feature but since we run all of them on the same env. it looks like we transfer the same data back and forth without any meaning. In its current state. But might be helpful after refactoring you suggest. Thanks.

**Comment by mitschabaude** [comment]
> > @shimkiv you could pass in the additional environment data like the CPU to `Benchmark.run()`, so that the benchmarking lib doesn't have to collect that data internally:
> 
> This is cool feature but since we run all of them on the same env. it looks like we transfer the same data back and forth without any meaning. In its current state. But might be helpful after refactoring you suggest. Thanks.

I agree, it's just passing data through. The alternative is to add the environment data to the benchmark results after collecting them.

In any case, the data needs to come from outside somehow if we go with the requirement that the benchmarking lib doesn't collect that data itself

**Comment by shimkiv** [comment]
@mitschabaude can you please take another look on latest refactoring in attempt to reflect what was requested (from my understanding).
We can also implement the file-system historical data preservation for local runs, but probably it should be done later in another PR.

**Review by ymekuria — COMMENTED** [review]
Nice work @shimkiv!  I will leave it up to @mitschabaude to decide if theses changes addressed his comments.

**Review by mitschabaude — COMMENTED** [review]
some comments left, looks great overall!

File: benchmark/samples/ecdsa.ts
[the remark this answers is not in the document]
Comment (shimkiv) [reply benchmark/samples/ecdsa.ts:1]: "samples", "test cases" or "benchmarks" all can be used interchangeably. I decided to use "samples" to not reduce tautology in dir and file names. 
Although don't have personal preferences.

**Review by mitschabaude — APPROVED** [review]
ship it 🚢

**Review by ymekuria — APPROVED** [review]

**Pull request merged**